refactor: centralize action buttons state

abeatrix opened the pull request:
Why: unify and simplify button behavior across task lifecycle, reduce duplicated state/props, and make streaming/approval flows more predictable.

<!--
Thank you for contributing to Cline!

⚠️ Important: Before submitting this PR, please ensure you have:
- For feature requests: Created a discussion in our Feature Requests discussions board https://github.com/cline/cline/discussions/categories/feature-requests and received approval from core maintainers before implementation
- For all changes: Link the associated issue/discussion in the "Related Issue" section below

Limited exceptions:
Small bug fixes, typo corrections, minor wording improvements, or simple type fixes that don't change functionality may be submitted directly without prior discussion.

Why this requirement?
We deeply appreciate all community contributions - they are essential to Cline's success! To ensure the best use of everyone's time and maintain project direction, we use our Feature Requests discussions board to gauge community interest and validate feature ideas before implementation begins. This helps us focus development efforts on features that will benefit the most users.
-->

### Related Issue

<!-- Replace XXXX with the issue number that this PR addresses -->
**Issue:** #XXXX

### Description

<!-- 
Help reviewers understand your changes by making this PR readable and well-organized:

- What problem does this PR solve?
- Why were these changes introduced and what purpose do they serve?
- For larger changes, provide context about your approach and reasoning

Small PRs may need minimal description, but larger changes benefit from explaining where you're coming from. Much of this context can be in the linked issue above, so feel free to reference it rather than repeating everything here.
-->

- Replace useButtonState hook with centralized ButtonConfig logic in ActionButtons, mapping task/ask/tool states to button enablement and labels
- Update ActionButtons API to accept task, messages, mode; compute streaming/enablement internally; remove isStreaming prop
- Always render ActionButtons from ChatView; adjust props accordingly
- Remove useIsStreaming hook that was used by the useButtonState hook which also depends on the useIsStreaming hook 
- Clean up useMessageHandlers to reset UI state consistently (input, quotes, files, images, autoscroll)
- Remove deprecated hook and align types

#### Problem

The chat interface had several critical button state synchronization issues:

1. Stale Button Text: Buttons would show outdated text like "Proceed While Running" even after the underlying state changed (e.g., when auto_approval_max_req_reached appeared)
2. Inconsistent State Management: Button state was managed in multiple places (useButtonState, useIsStreaming, ActionButtons) leading to race conditions and inconsistent updates
3. Timing Issues: Button states weren't updating in sync with message changes, causing UI to show incorrect actions to users

Example: we are showing disabled buttons after the model didn't provide any response:

1. All tasks start with the "Approve" and "Reject" buttons with them disabled: https://github.com/cline/cline/blob/main/webview-ui/src/components/chat/chat-view/hooks/useButtonState.ts#L152-L159
2. We just don't show it until the streaming has started: https://github.com/cline/cline/blob/main/webview-ui/src/components/chat/chat-view/components/layout/ActionButtons.tsx#L54
3. But once the streaming has started and there was no response, we will show the buttons that we last set to, in this case it's the original state
4. The useIsStreaming hook and useButtonState hook also have shared dependencies to determine if a task has partial response or not, but they were also checking different task state (one is tracking last message, the other one is tracking modified message that removes the last message). Then the current isStreaming value comes from the useIsStreaming hook that also depends on the enableButtons & primaryButtonText values which depends on the isStreaming value, which 
I think this is the root cause to a lot of the issues where users got locked out of their current task

##### Solution

1. Centralized Button Configuration

- Created buttonConfig.ts as single source of truth for all button states and actions
- Defined ButtonActionType enum to standardize button behaviors
- Consolidated all button configurations into BUTTON_CONFIGS with clear state mappings

2. Eliminated State Management Hooks

- Removed useButtonState hook that was causing stale state issues
- Moved useIsStreaming logic directly into ActionButtons component for real-time updates
- Simplified state management by computing button states directly from current message

3. Unified Button Action Handling

- Replaced separate handlePrimaryButtonClick/handleSecondaryButtonClick with single handleButtonClick(action, ...)
- Actions are now determined by button text mapping to ButtonActionType, ensuring consistency
- Eliminated the disconnect between what buttons show and what they actually do

5. Real-time State Updates

- Button states now update immediately when messages change via useMemo and useEffect
- Special handling for edge cases like api_req_started after command_output
- Proper cleanup of input state when transitioning between different ask types

6. Improved Component Architecture

- ActionButtons now receives task and messages directly for real-time state computation
- Removed dependency on separate streaming state hooks
- Better separation of concerns between state computation and UI rendering


### Test Procedure

<!-- 
Please walk us through your testing approach and thought process. This helps reviewers understand that you've thoroughly considered the impact of your changes:

- How did you test this change?
- What could potentially break and how did you verify it doesn't?
- What existing functionality might be affected and how did you check it still works?
- Why are you confident this is ready for merge?

We're not looking for exhaustive documentation - just evidence that you've thought through the implications of your changes and tested accordingly.
-->

1. Buttons now immediately reflect the correct state when messages change
2. No more "Proceed While Running" stuck on screen when `auto_approval_max_req_reached` appears
3. Button actions consistently match their displayed text
5. Streaming states properly disable/enable buttons in real-time

This refactor ensures button states are always synchronized with the actual conversation state, eliminating the confusing UX where buttons showed one action but would perform another.

### Type of Change

<!-- Put an 'x' in all boxes that apply -->

-   [x] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [ ] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [x] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

<!-- Put an 'x' in all boxes that apply -->

-   [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [x] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
-   [ ] I have created a changeset using `npm run changeset` (required for user-facing changes)
-   [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Screenshots

<!-- 
Help reviewers quickly understand your changes:

- **UI Changes**: Please include screenshots showing before/after states
- **Complex Workflows**: Consider uploading a screen recording (video) if your changes involve multiple steps or state transitions
- **Backend Changes**: Not required, but feel free to include terminal output or other evidence that demonstrates functionality

This helps reviewers see what you've built without having to pull down and test your branch first.
-->

<img width="769" height="790" alt="image" src="https://github.com/user-attachments/assets/cb1b2861-f055-4b23-b2b0-39f34b45568c" />

<img width="1771" height="1014" alt="image" src="https://github.com/user-attachments/assets/530dc5e1-175c-495d-8927-e2a38cfeff29" />

### Additional Notes

<!-- Add any additional notes for reviewers -->

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Centralizes button state management in chat interface, updating `ActionButtons` and `useMessageHandlers` to use new configuration logic, and adds tests for button states.
> 
>   - **Behavior**:
>     - Centralizes button state management in `buttonConfig.ts`, defining `BUTTON_CONFIGS` and `getButtonConfig()`.
>     - Updates `ActionButtons` to use centralized button logic, removing `isStreaming` prop.
>     - Removes `useButtonState` and `useIsStreaming` hooks, integrating logic into `ActionButtons`.
>   - **Functions**:
>     - Updates `useMessageHandlers` to handle button actions using `ButtonActionType`.
>     - Adds `handleButtonClick()` in `useMessageHandlers` for unified button action handling.
>   - **Tests**:
>     - Adds `buttonConfig.test.ts` to test button configurations for various states.
>   - **Misc**:
>     - Adjusts `ChatView` to always render `ActionButtons` and update props accordingly.
>     - Cleans up `useChatState` to remove unused state management.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 112f11f8cec04dff5a2c0350105e685b9815c88d. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

abeatrix commented on webview-ui/src/components/chat/chat-view/hooks/useMessageHandlers.ts:
```diff
@@ -129,175 +130,152 @@ export function useMessageHandlers(messages: ClineMessage[], chatState: ChatStat
+			setDidClickCancel,
 			setSendingDisabled,
 			setEnableButtons,
 			chatState,
```
chatState is a hooke

abeatrix commented on webview-ui/src/components/chat/chat-view/shared/buttonConfig.ts:
```diff
@@ -0,0 +1,310 @@
+	},
+}
+
+const errorTypes = ["api_req_failed", "mistake_limit_reached", "auto_approval_max_req_reached"]
```
i think it's pretty self explanatory already

arafatkatze reviewed (changes requested):
There are different kinds of state transitions here that would all need to be tested and I will do that.

One case that comes to mind is that if I ever stop mid streaming it shows start new task instead of resume task(like before. )

arafatkatze commented on webview-ui/src/components/chat/chat-view/components/messages/StreamingIndicator.tsx:
```diff
@@ -1,22 +1,21 @@
-	enableButtons?: boolean,
-	primaryButtonText?: string,
-): boolean => {
+export const useIsStreaming = (modifiedMessages: ClineMessage[], clineAsk?: string, task?: ClineMessage): boolean => {
```
This is NOT being called anywhere in the codebase.

abeatrix replied:
deleted

arafatkatze reviewed (commented):
<img width="454" height="231" alt="image" src="https://github.com/user-attachments/assets/952c1bf8-5fde-46d5-bd1d-bc8a90015783" />
Can this be resolved

abeatrix commented on webview-ui/src/components/chat/chat-view/components/layout/ActionButtons.tsx:
```diff
@@ -17,26 +22,98 @@ interface ActionButtonsProps {
+		}
+
+		return getButtonConfig(lastMessage, mode)
+	}, [chatState, lastMessage, secondLastMessage, mode])
```
moved the state updating logic to useEffect

abeatrix commented:
> One case that comes to mind is that if I ever stop mid streaming it shows start new task instead of resume task(like before. )
That's because I updated the button text , which I reverted.

arafatkatze commented on webview-ui/src/components/chat/chat-view/components/layout/ActionButtons.tsx:
```diff
@@ -17,26 +22,107 @@ interface ActionButtonsProps {
+			}
+			if (chatState.selectedFiles.length) {
+				chatState.setSelectedFiles([])
+			}
```
Can you also clear the cancel 
<img width="364" height="282" alt="image" src="https://github.com/user-attachments/assets/edb35901-d365-482e-acf9-764ccf3f31b0" />

arafatkatze replied:
So when you start a new request this should be cleared. Because once you set it to true when do we cancel it?

abeatrix replied:
thanks for clarifying! I think this shows the issue with how we are managing states in different places that show be unified into a centralized place! I have removed the `setDidClickCancel`  altogether.

arafatkatze reviewed (approved):
Tested the PR in all different ways that I could think off and for me it worked fine and feature parity with the previous use cases has been maintained. 

I am thankful for this great improvement to the codebase and I hope this solves all different kinds of state synchronization issues with the action button.

It would have been very nice to also fix some state transitions within this PR alone but just to be extra cautious that's not being done so that we can maintain feature parity and get this out in front of people and if no one complains in the subsequent release we can have state transitions to make more sense. 

Baked into the codebase we have a couple of bad state transitions like this one 

<img width="640" height="720" alt="image" src="https://github.com/user-attachments/assets/3069ef1f-92ba-436a-bd64-ada1428de0b1" />


But its better to just wait and let people try things out with feature parity from the past and then do more changes in the followup given that this is very sensitive code.

arafatkatze commented:
🙇
